letsencrypt/hooks.py

```diff
@@ -0,0 +1,99 @@
+"""Facilities for implementing hooks that call shell commands."""
+from __future__ import print_function
+
+import logging
+import os
+
+from subprocess import Popen, PIPE
+
+from letsencrypt import errors
+
+logger = logging.getLogger(__name__)
+
+def validate_hooks(config):
+    """Check hook commands are executable."""
+    _validate_hook(config.pre_hook, "pre")
+    _validate_hook(config.post_hook, "post")
+    _validate_hook(config.renew_hook, "renew")
+
+def _prog(shell_cmd):
+    """Extract the program run by a shell command"""
+    cmd = _which(shell_cmd)
+    return os.path.basename(cmd) if cmd else None
+
+def _validate_hook(shell_cmd, hook_name):
+    """Check that a command provided as a hook is plausibly executable.
+
+    :raises .errors.HookCommandNotFound: if the command is not found
+    """
+    cmd = shell_cmd.partition(" ")[0]
+    if shell_cmd and not _prog(cmd):
+        path = os.environ["PATH"]
+        msg = "Unable to find {2}-hook command {0} in the PATH.\n(PATH is {1})".format(
+            cmd, path, hook_name)
+        raise errors.HookCommandNotFound(msg)
+
+    return True
+
+def pre_hook(config):
+    "Run pre-hook if it's defined and hasn't been run."
+    if config.pre_hook and not pre_hook.already:
+        logger.info("Running pre-hook command: %s", config.pre_hook)
+        _run_hook(config.pre_hook)
+        pre_hook.already = True
+
+pre_hook.already = False
+
+def post_hook(config, final=False):
+    """Run post hook if defined.
+
+    If the verb is renew, we might have more certs to renew, so we wait until
+    we're called with final=True before actually doing anything.
+    """
+    if config.post_hook:
+        if final or config.verb != "renew":
+            logger.info("Running post-hook command: %s", config.post_hook)
+            _run_hook(config.post_hook)
```

Comment:
If I were writing a post-hook script, I can imagine that I would want more programmatic access to information about what happened, for example a `RENEW_SUCCESS` variable or something that indicates whether renewals succeeded. It's a little hard for me to make super-specific suggestions about what information to provide because I haven't had any actual need to write these scripts, but maybe we should ask on the forum?